### PR TITLE
nix-prefetch-scripts: Add prefetcher for github

### PR DESCRIPTION
**Pull request by avnik**
Implements prefetchers for major source hostings, corresponding to
`fetchFromGitHub`, `fetchFromBitbucket` and others.

### CODE REVIEW

**Comment by trishume** [comment]
Great idea!

it would be even cooler if you could paste the url of a Github project after the command and it would parse out the repo and user, and fetch and print the SHA of the latest revision, or maybe even print the `fetchFromGithub` block that you can copy-paste into your config file.

**Comment by matthiasbeyer** [comment]
Please do not merge. I want to provide some patches to this.

File: pkgs/tools/package-management/nix-prefetch-scripts/nix-prefetch-github [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+              usage
+              ;;
+         esac
+         ;;
+    esac
+  else
+    case $argfun in
+      set_*)
+        var=$(echo $argfun | sed 's,^set_,,')
+        eval "$var=\$arg"
```
Comment (matthiasbeyer) [inline pkgs/tools/package-management/nix-prefetch-scripts/nix-prefetch-github]: Holy shit, you're doing `eval` here.

*PLEASE* do not merge this. I'm sure I can make this less dangerous!

**Comment by matthiasbeyer** [comment]
Added PR to this PR in https://github.com/avnik/nixpkgs/pull/1.

**Comment by jagajaga** [comment]
I can't understand what is it for? :)

**Comment by matthiasbeyer** [comment]
@jagajaga `nix-prefetch-github --owner jagajaga --repo myPackage --rev v0.1` vs. `nix-prefetch-git https://github.com/jagajaga/myPackage/archive/v0.1.tar.gz`

That's it I guess. Anyways, with my patches from linked PR to this PR, we could build a script where you can pass `github.com/jagajaga/myPackage` and `v0.1` (or with the more complex urls from savannah) and it automatically finds out what you're trying to download.

But yeah, I don't know whether we want something like that. Pasting in the URL is propably enough. Maybe @avnik can explain? :smile:

**Comment by jagajaga** [comment]
For `nix-prefetch-git https://github.com/jagajaga/myPackage/archive/v0.1.tar.gz` we have `nix-prefetch-url`. For git repos we have `nix-prefetch-git`. I think this is not so much useful.

**Comment by trishume** [comment]
Good point @jagajaga, it's actually easier to right-click "copy URL" on the archive download on Github then it is to use this script.

**Comment by Mathnerd314** [comment]
nix-prefetch-url does not work here, it generates the wrong hash (because fetchFromGithub uses fetchzip & the hash is of the unpacked contents). And nix-prefetch-zip does not work because it generates the wrong name from the URL.

**Comment by jagajaga** [comment]
cc @edolstra

**Comment by Mathnerd314** [comment]
And note that although nix-prefetch-git generates the correct hash, it does not generate the correct output path, so it does not work for large git repositories that you want to only download once.

Here is my testing on a small repo:
```
nix-prefetch-git: /nix/store/0qrzq7ikblxg9fnqv1g5y7h69zbsp2n2-git-export
nix-prefetch-url: /nix/store/yj2bqq28s2a8fw7n3dkbk55lylvibmz2-6cbd7c5dcdad5beee3a870e60cacaf0a3a1b0dee.tar.gz
nix-prefetch-zip: /nix/store/0912hgki3fn12yyj4aahv6jnsbz5bqvm-6cbd7c5dcdad5beee3a870e60cacaf0a3a1b0dee.tar.gz
 fetchFromGitHub: /nix/store/j4cmz4jin1y60cg6zjav7qzj00rqmr79-cq-challenge-markup-6cbd7c5dcdad5beee3a870e60cacaf0a3a1b0dee-src
```

**Comment by jagajaga** [comment]
@Mathnerd314 so fix `nix-prefetch-git`? It's not good to develop new complicated thing if we can improve the existing one.

**Comment by matthiasbeyer** [comment]
I'd like to suggest to close this unmerged. I think there is no point in such commands. Sorry OP!

**Comment by jagajaga** [comment]
@matthiasbeyer :+1:
Will close in a day if no objections.

File: pkgs/tools/package-management/nix-prefetch-scripts/nix-prefetch-github (hunk 1 above)
Comment (avnik) [inline pkgs/tools/package-management/nix-prefetch-scripts/nix-prefetch-github]: Parsing of commandline done in all other tools exactly in same style (this fragment copied from nix-prefetch-zip)

**Comment by avnik** [comment]
@jagajaga This set of scripts just wrappers over `nix-prefetch-zip` (build proper url from owner/repo/rev and pass it to nix-prefetch-url

PS Sorry for my late appearance in thread -- mail lagging

**Comment by trishume** [comment]
@avnik I think what @jagajaga seems to be getting at is that instead of wrapping the interface to fix the issue maybe you should just add the handling code  (maybe as an option) to either `nix-prefetch-zip` or `nix-prefetch-git` where you give it a URL like `nix-prefetch-zip https://github.com/jagajaga/myPackage/archive/v0.1.tar.gz`

That way is actually simpler to use, and it doesn't add more binaries to the NixOS distribution. Also less code since most of this PR is actually argument parsing code that just passes the arguments on to `nix-prefetch-zip`

**Comment by jagajaga** [comment]
@trishume is right. Less code = easier to maintain it.

**Comment by avnik** [comment]
On Wed, Sep 23, 2015 at 09:31:59AM -0700, Tristan Hume wrote:
> @avnik I think what @jagajaga seems to be getting at is that 
> instead of wrapping the interface to fix the issue maybe you
> should just add the handling code  (maybe as an option) 
> to either `nix-prefetch-zip` or `nix-prefetch-git` where you
> give it a URL like `nix-prefetch-zip 
> https://github.com/jagajaga/myPackage/archive/v0.1.tar.gz`
> 

Well, we have a lot of option.
* Add option to `nix-prefetch-git` to switch to `nix-prefetch-zip`,
in case if we want to fetch from github/gitlab/other place which
provide snapshots.
* May be add some extra options and allow `nix-prefetch-zip` to 
form urls from --site/--owner/--rev/...


> That way is actually simpler to use, and it doesn't add more binaries to the NixOS distribution. Also less code since most of this PR is actually argument parsing code that just passes the arguments on to `nix-prefetch-zip`
> 

Want to keep this PR for discuss, then will close it myself.

**Comment by edolstra** [comment]
I would like to propose:
* Fold `nix-prefetch-zip` into `nix-prefetch-url`.
* Extend `nix-prefetch-url` to fetch from a Nix expression, e.g. `nix-prefetch-url -A firefox.src`. This would then work for all wrappers around `fetchurl` and `fetchzip` (like `fetchFromGitHub`) automatically.

**Comment by avnik** [comment]
On Thu, Sep 24, 2015 at 06:11:41AM -0700, Eelco Dolstra wrote:

> I would like to propose:
> * Fold `nix-prefetch-zip` into `nix-prefetch-url`.
> * Extend `nix-prefetch-url` to fetch from a Nix expression, e.g. `nix-prefetch-url -A firefox.src`. This would then work for all wrappers around `fetchurl` and `fetchzip` (like `fetchFromGitHub`) automatically.
> 

Would "fetch from nix expression" thing work with stuff with 
not-yet-known hashes?

**Comment by jagajaga** [comment]
Any updates, question, suggestions? What should we do? :)

**Comment by Mathnerd314** [comment]
@edolstra  has implemented https://github.com/NixOS/nix/tree/master/src/nix-prefetch-url, I'm not sure of its completeness

**Comment by joachifm** [comment]
Seems to me the consensus here is against this PR. Closing.

**Pull request closed without merging**